### PR TITLE
[patch] Fix cache bug

### PR DESCRIPTION
**Pull request by ssunorz**
# Description

_description_with_details_and_reasoning_

## Type of change

- [x] Bug fix
- [ ] New feature
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Non-code changes (update documentation, pipeline, etc.)

### Flags

- [ ] breaks backward compatibility
- [ ] requires a documentation update
- [ ] has untestable code

## Related issue/PR

**Delete if not applicable**
- Fixes #_issue_
- Closes #_PR_

---

## Checklist

- [ ] Followed the guidelines in the CONTRIBUTING document
- [ ] Added prefix `[major]`/`[minor]`/`[patch]`/`[skip]` in the PR title
- [ ] Tested and linted the code
- [ ] Commented the code
- [ ] Made corresponding changes to the documentation
- [ ] Confirmed no dropping in test coverage (by [Codecov](https://codecov.io/gh/yahoojapan/athenz-authorizer/pulls))
- [ ] Passed all pipeline checking
- [ ] Approved by >1 reviewer

## Checklist for maintainer
- [ ] Use `Squash and merge`
- [ ] Double-confirm the merge message has prefix `[major]`/`[minor]`/`[patch]`/`[skip]`
- [ ] Delete the branch after merge

### CODE REVIEW

File: authorizerd.go [hunk 1]
```diff
@@ -356,13 +356,17 @@ func (a *authority) AuthorizeAccessToken(ctx context.Context, tok, act, res stri
...
 	if a.disablePolicyd {
-		key = tok
+		key = key + tok
 	} else {
 		if act == "" || res == "" {
 			return nil, errors.Wrap(ErrInvalidParameters, "empty action / resource")
 		}
-		key = tok + act + res
+		key = key + tok + act + res
 	}
```
Comment (WindzCUHK) [inline authorizerd.go:370]: 1. how about using `strings.Builder`?
   - https://stackoverflow.com/questions/1760757/how-to-efficiently-concatenate-strings-in-go/47798475#47798475
2. add delimiter between all sub-strings

---

1. var key strings.Builder
2. key.WriteString(tok); key. WriteRune(':');
3. if cert != nil, key.WriteString(common_name); key. WriteRune(':');
4. if !disablePolicyd, key.WriteString(act); key. WriteRune(':'); key. WriteString(res);

Comment (ssunorz) [reply authorizerd.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/75/commits/ccdbdd41c43685852a6d6e66cd1fe1ed9c8e0533

File: authorizerd.go [hunk 2]
```diff
@@ -354,21 +354,30 @@ func (a *authority) AuthorizeAccessToken(ctx context.Context, tok, act, res stri
 }
 
 func (a *authority) authorize(ctx context.Context, m mode, tok, act, res string, cert *x509.Certificate) (Principal, error) {
-	var key string
+	var key strings.Builder
+	key.WriteString(tok)
+
+	if cert != nil {
+		key.WriteRune(':')
```
Comment (WindzCUHK) [inline authorizerd.go:361]: Better to make the `':'` a constant/variable?

Comment (ssunorz) [reply authorizerd.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/75/commits/2f4da3ffc4091a1395b41a8d5c53b87131c92597

File: authorizerd_test.go [hunk 3]
```diff
@@ -1614,10 +1686,75 @@ func Test_authorizer_AuthorizeAccessToken(t *testing.T) {
...
+					policyd:         pdm,
+					accessProcessor: apm,
+					cache:           c,
+					cacheExp:        time.Minute,
+				},
+				wantErr:    "",
+				wantResult: p,
+				checkFunc: func(prov *authority) error {
+					_, expiry, ok := prov.cache.GetWithExpire("dummyTok:issuer cn:subject cn:dummyAct:dummyRes")
+					if !ok {
```
Comment (kyfujisa) [inline authorizerd_test.go:1742]: If you want to test that the cache is used, then how about checking the length of the cache?
```suggestion
					if !ok || prov.cache.Len() != 1 {
```
If you agree, please make the same change to line 1677.

Comment (ssunorz) [reply authorizerd_test.go]: Fixed
https://github.com/yahoojapan/athenz-authorizer/pull/75/commits/d66340eeb7d1b116772ec9d30dd98d5a10fac43a

**Pull request merged**